src/python/pants/backend/native/subsystems/native_toolchain.py

```diff
@@ -63,211 +67,226 @@ def _libc_dev(self):
-  # and C++ compilers baked in.
-  # FIXME(#5951): we need a way to compose executables more hygienically.
-  linker = Linker(
-    path_entries=(
-      cpp_compiler.path_entries +
-      c_compiler.path_entries +
-      linker.path_entries),
-    exe_filename=cpp_compiler.exe_filename,
-    library_dirs=(
-      libc_dirs +
-      cpp_compiler.library_dirs +
-      c_compiler.library_dirs +
-      linker.library_dirs))
-
-  yield linker
-
-
-@rule(LLVMCCompiler, [Select(Platform), Select(NativeToolchain)])
-def select_llvm_c_compiler(platform, native_toolchain):
+  base_linker = BaseLinker(linker=linker)
+  yield base_linker
+
+
+@rule(LLVMCToolchain, [Select(Platform), Select(NativeToolchain)])
+def select_llvm_c_toolchain(platform, native_toolchain):
   original_llvm_c_compiler = yield Get(LLVMCCompiler, LLVM, native_toolchain._llvm)
   provided_clang = original_llvm_c_compiler.c_compiler
 
+  # These arguments are shared across platforms.
+  llvm_c_compiler_args = [
+    '-x', 'c', '-std=c11',
+    # These mean we don't use any of the headers from our LLVM distribution.
+    '-nobuiltininc',
+  ]
+
   if platform.normalized_os_name == 'darwin':
     xcode_clang = yield Get(CCompiler, XCodeCLITools, native_toolchain._xcode_cli_tools)
-    clang_with_xcode_paths = CCompiler(
+    working_c_compiler = CCompiler(
       path_entries=(provided_clang.path_entries + xcode_clang.path_entries),
       exe_filename=provided_clang.exe_filename,
       library_dirs=(provided_clang.library_dirs + xcode_clang.library_dirs),
-      include_dirs=(xcode_clang.include_dirs + provided_clang.include_dirs))
-    final_llvm_c_compiler = LLVMCCompiler(clang_with_xcode_paths)
+      include_dirs=(provided_clang.include_dirs + xcode_clang.include_dirs),
+      extra_args=(llvm_c_compiler_args + xcode_clang.extra_args))
   else:
     gcc_c_compiler = yield Get(GCCCCompiler, GCC, native_toolchain._gcc)
     provided_gcc = gcc_c_compiler.c_compiler
-    clang_with_gcc_libs = CCompiler(
+    working_c_compiler = CCompiler(
       path_entries=provided_clang.path_entries,
       exe_filename=provided_clang.exe_filename,
-      # We need this version of GLIBCXX to be able to run, unfortunately.
+      # We need g++'s version of the GLIBCXX library to be able to run, unfortunately.
       library_dirs=(provided_gcc.library_dirs + provided_clang.library_dirs),
-      include_dirs=(provided_clang.include_dirs + provided_gcc.include_dirs))
-    final_llvm_c_compiler = LLVMCCompiler(clang_with_gcc_libs)
+      include_dirs=provided_gcc.include_dirs,
+      extra_args=llvm_c_compiler_args)
 
-  yield final_llvm_c_compiler
+  base_linker_wrapper = yield Get(BaseLinker, NativeToolchain, native_toolchain)
+  base_linker = base_linker_wrapper.linker
+  libc_dev = yield Get(LibcDev, NativeToolchain, native_toolchain)
+  working_linker = Linker(
+    path_entries=(base_linker.path_entries + working_c_compiler.path_entries),
+    exe_filename=working_c_compiler.exe_filename,
```

Comment:
So we are still linking through the compiler frontend? Based on `path_entries`, it looks like the we will hit our linker first, however why not explicitly invoke it? I know you explained this to me once but I'm having a tough time recalling...I think it had to do with critical search dirs that we are unable to find by directly executing the linker?

Reply:
It is almost literally impossible to invoke the linker directly. If you try adding `-v` to the `extra_args` for a `Linker` somewhere you can see the generated command line -- I have not been able to make it work yet, after trying for a while, and as far as I can tell there is no reason we would to invoke the linker directly, unless the compiler frontend is pulling in something we don't want (which is what we are using the `extra_args` to avoid, with arguments such as `-nostdinc++`).

Reply:
(to clarify, I would love to invoke the linker directly, but it's not something that has ever borne fruit every time I've tried. if there's a way to do this that I'm missing I'm all ears)

Reply:
(what I said to you before was a condensed version of that)

Reply:
Sure, I don't think this is a problem if we are hitting our linker regardless. If we find that the frontend is making things too difficult to build/debug, we can lean into direct invocation.